test/service_latency_test.go

```diff
@@ -0,0 +1,439 @@
+	})
+
+	// Same Cluster Requestor
+	nc2 := clientConnect(t, sc.clusters[0].opts[2], "bar")
+	defer nc.Close()
+
+	toSend := 1000
+	for i := 0; i < toSend; i++ {
+		nc2.Request("ngs.usage", []byte("1h"), time.Second)
+	}
+	// Wait for results to flow in.
+	time.Sleep(100 * time.Millisecond)
+
+	mid := toSend / 2
+	delta := toSend / 10 // 10%
+	got := int(atomic.LoadInt32(&received))
+
+	if got > mid+delta || got < mid-delta {
+		t.Fatalf("Sampling number incorrect: %d vs %d", mid, got)
+	}
+}
+
+func TestServiceLatencyWithName(t *testing.T) {
+	sc := createSuperCluster(t, 1, 1)
+	defer sc.shutdown()
+
+	// Now add in new service export to FOO and have bar import that with tracking enabled.
+	sc.setupLatencyTracking(t, 100)
+
+	opts := sc.clusters[0].opts[0]
+
+	nc := clientConnectWithName(t, opts, "foo", "dlc22")
+	defer nc.Close()
+
+	// The service listener.
+	nc.Subscribe("ngs.usage.*", func(msg *nats.Msg) {
+		msg.Respond([]byte("22 msgs"))
+	})
+
+	// Listen for metrics
+	rsub, _ := nc.SubscribeSync("results")
+
+	nc2 := clientConnect(t, opts, "bar")
+	defer nc.Close()
+	nc2.Request("ngs.usage", []byte("1h"), time.Second)
+
+	var sl server.ServiceLatency
+	rmsg, _ := rsub.NextMsg(time.Second)
+	json.Unmarshal(rmsg.Data, &sl)
+
+	// Make sure we have AppName set.
+	if sl.AppName != "dlc22" {
+		t.Fatalf("Expected to have AppName set correctly, %q vs %q", "dlc22", sl.AppName)
+	}
+}
+
+func TestServiceLatencyWithNameMultiServer(t *testing.T) {
+	sc := createSuperCluster(t, 3, 2)
+	defer sc.shutdown()
+
+	// Now add in new service export to FOO and have bar import that with tracking enabled.
+	sc.setupLatencyTracking(t, 100)
+
+	nc := clientConnectWithName(t, sc.clusters[0].opts[1], "foo", "dlc22")
+	defer nc.Close()
+
+	// The service listener.
+	nc.Subscribe("ngs.usage.*", func(msg *nats.Msg) {
+		msg.Respond([]byte("22 msgs"))
+	})
+
+	// Listen for metrics
+	rsub, _ := nc.SubscribeSync("results")
+
+	nc2 := clientConnect(t, sc.clusters[1].opts[1], "bar")
+	defer nc.Close()
+	nc2.Request("ngs.usage", []byte("1h"), time.Second)
+
+	var sl server.ServiceLatency
+	rmsg, _ := rsub.NextMsg(time.Second)
+	json.Unmarshal(rmsg.Data, &sl)
+
+	// Make sure we have AppName set.
+	if sl.AppName != "dlc22" {
+		t.Fatalf("Expected to have AppName set correctly, %q vs %q", "dlc22", sl.AppName)
+	}
+}
+
+func TestServiceLatencyWithQueueSubscribersAndNames(t *testing.T) {
+	numServers := 3
+	numClusters := 3
+	sc := createSuperCluster(t, numServers, numClusters)
+	defer sc.shutdown()
+
+	// Now add in new service export to FOO and have bar import that with tracking enabled.
+	sc.setupLatencyTracking(t, 100)
+
+	selectServer := func() *server.Options {
+		si, ci := rand.Int63n(int64(numServers)), rand.Int63n(int64(numServers))
+		return sc.clusters[ci].opts[si]
+	}
+
+	sname := func(i int) string {
+		return fmt.Sprintf("SERVICE-%d", i+1)
+	}
+
+	// Create 10 queue subscribers for the service. Randomly select the server.
+	for i := 0; i < 10; i++ {
```

Comment:
I have ran this locally and it is going to be flaky with 10 qsubs and only 100 requests. Since we artificially sleep, and don't want the test to take too long, we can't increase the number of requests, but should reduce the number of qsubs to 5 to avoid situations where some of the qsubs do not receive any requests or only 1 or 2 and therefore possibly are below the 10ms threshold.

Reply:
Will lower number of responders.